src/Microsoft.OData.Core/JsonLight/ODataJsonLightOutputContext.cs

```diff
@@ -953,32 +945,73 @@ protected override void Dispose(bool disposing)
+            Debug.Assert(asynchronousJsonWriter != null, "asynchronousJsonWriter != null");
+
+            if (asynchronousJsonWriter is Json.JsonWriter defaultJsonWriter)
+            {
+                defaultJsonWriter.ArrayPool = writerSettings.ArrayPool;
             }
-            else
+
+            return asynchronousJsonWriter;
+        }
+
+        private static IJsonWriterAsync CreateAsynchronousJsonWriter(
+            IStreamBasedJsonWriterFactory factory,
+            Stream outputStream,
+            bool isIeee754Compatible,
+            Encoding encoding)
+        {
+            IJsonWriterAsync jsonWriter = factory.CreateAsynchronousJsonWriter(outputStream, isIeee754Compatible, encoding);
+
+            if (jsonWriter == null)
+            {
+                throw new ODataException(Strings.ODataMessageWriter_StreamBasedJsonWriterFactory_ReturnedNull(encoding.WebName, isIeee754Compatible));
+            }
+
+            return jsonWriter;
+        }
+
+        private void VerifyJsonWriter()
+        {
+            // Asynchronous support is not implemented in Microsoft.Spatial library.
+            // To write spatial data, we rely on the synchronous PrimitiveConverter.Instance.WriteJsonLight(object, IJsonWriter) method.
+            // When writing asynchronously we wrap this method in a Task. WriteJsonLight method takes an
+            // IJsonWriter parameter while the asynchronous writer is declared as IJsonWriterAsync.
+            // When writing asynchronously, we have to ensure that the IJsonWriter that is used
+            // for writing spatial data is the same instance as the IJsonWriterAsync used for writing
+            // everything else in order to guarantee that the writer state is correctly maintained
+            // throughout the writing process (e.g. keeping track of the writer's current scope).
+            // When an IJsonWriterAsync is provided, it must also implement IJsonWriter so that the same instance
+            // can be reused for spatial data. If we somehow end up with 2 separate instances, then we fail early with an exception.
+            // Merging IJsonWriter and IJsonWriterAsync interface in a major release will simplify this.
+
+            if (this.jsonWriter != null && this.asynchronousJsonWriter == null && this.jsonWriter is IJsonWriterAsync jsonWriterAsync)
+            {
+                this.asynchronousJsonWriter = jsonWriterAsync;
+                return;
+            }
+
+            if (this.asynchronousJsonWriter != null && this.jsonWriter == null)
             {
-                IJsonWriterFactoryAsync asynchronousJsonWriterFactory = container.GetRequiredService<IJsonWriterFactoryAsync>();
-                if (asynchronousJsonWriterFactory is DefaultJsonWriterFactory && this.concreteJsonWriter != null)
+                if (this.asynchronousJsonWriter is IJsonWriter syncJsonWriter)

```

Comment:
why didn't you combine the two statements here like the `if` above:

`if (this.asynchronousJsonWriter != null && this.jsonWriter == null && this.asynchronousJsonWriter is IJsonWriter syncJsonWriter)`

Reply:
Can it be documented so they know that they can use Async Json Writer if they do not call the `AddDefaultODataServices()`

Reply:
Yes, will mention that when updating the docs.